android/src/main/java/com/theoplayer/presentation/PresentationManager.kt

```diff
@@ -209,50 +217,63 @@ class PresentationManager(
     val activity = reactContext.currentActivity ?: return
     val window = activity.window
 
-    // Get the player's ReactViewGroup parent, which contains THEOplayerView and its children (typically the UI).
-    val reactPlayerGroup: ReactViewGroup? = getClosestParentOfType(this.viewCtx.playerView)
-
-    // Get ReactNative's root node or the render hierarchy
-    val root: ReactRootView? = getClosestParentOfType(reactPlayerGroup)
-
     if (fullscreen) {
       WindowInsetsControllerCompat(window, window.decorView).apply {
         systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
       }.hide(WindowInsetsCompat.Type.systemBars())
       updatePresentationMode(PresentationMode.FULLSCREEN)
 
-      if (!BuildConfig.REPARENT_ON_FULLSCREEN) {
-        return
-      }
-      playerGroupParentNode = if (BuildConfig.IS_NEW_ARCHITECTURE_ENABLED) {
-        reactPlayerGroup?.parent as? ReactSurfaceView?
-      } else {
-        reactPlayerGroup?.parent as? ReactViewGroup?
-      }?.also { parent ->
-        playerGroupChildIndex = parent.indexOfChild(reactPlayerGroup)
-        // Re-parent the playerViewGroup to the root node
-        parent.removeView(reactPlayerGroup)
-        root?.addView(reactPlayerGroup)
+      if (BuildConfig.REPARENT_ON_FULLSCREEN) {
+        reparentPlayerToRoot()
       }
     } else {
       WindowInsetsControllerCompat(window, window.decorView).show(
         WindowInsetsCompat.Type.systemBars()
       )
       updatePresentationMode(PresentationMode.INLINE)
 
-      if (!BuildConfig.REPARENT_ON_FULLSCREEN) {
-        return
-      }
-      root?.run {
-        // Re-parent the playerViewGroup from the root node to its original parent
-        removeView(reactPlayerGroup)
-        playerGroupParentNode?.addView(reactPlayerGroup, playerGroupChildIndex ?: 0)
-        playerGroupParentNode = null
-        playerGroupChildIndex = null
+      if (BuildConfig.REPARENT_ON_FULLSCREEN) {
+        reparentPlayerToOriginal()
       }
     }
   }
 
+  // region Re-parent playerViewGroup logic
+  private val reactPlayerGroup: ReactViewGroup?
+    get() = viewCtx.playerView.getClosestParentOfType()
+  private val rootView: ReactRootView?
+    get() {
+      val activity = reactContext.currentActivity ?: return null
+      // Try to search in parents and as a fallback option from root to bottom using depth-first order
+      return reactPlayerGroup?.getClosestParentOfType()
+        ?: (activity.window.decorView.rootView as? ViewGroup)
+          ?.getClosestParentOfType(false)
```

Comment:
would work as a fallback option, so should be safe to merge for current clients  I guess. The part to test is PiP reparenting but we set the build flag  to false by default as it is experimental

Reply:
Just for my understanding: in which case did the upwards search for a suitable root fail?

Reply:
in our apps it happens when we switch on the fullscreen mode from inline player. For some reason the React Root view doesn't exist in the direct player's parents in this particular scenario. Also it happens only on the second and any subsequent times. The first attempt it just works fine with the upward search.